Update ref-struct-interfaces.md (#8024)
dotnet/csharplang: proposals/ref-struct-interfaces.md

ref struct S = I1 { }
```

To handle this a `ref struct` will be forced to implement all members of an interface, even if they have default implementations.
The runtime will also be updated to throw an exception if a default interface member is called on a `ref struct` type.

To avoid an exception at runtime the compiler will report an error for an invocation of a non-virtual instance method (or property)
on a type parameter that allows ref struct. Here is an example:
```csharp
public interface I1
{
    sealed void M3() {}
}

class C
{
    static void Test2<T>(T x) where T : I1, allows ref struct
    {
#line 100

>         x.M3(); // (100,9): error CS9506: A non-virtual instance interface member cannot be accessed on a type parameter that allows ref struct.

Reviewer: Consider removing the error code CS9506 if it's not final

>         x.M3(); // (100,9): error CS9506: A non-virtual instance interface member cannot be accessed on a type parameter that allows ref struct.

Author: ```suggestion
        x.M3(); // (100,9): error: A non-virtual instance interface member cannot be accessed on a type parameter that allows ref struct.
```

Whether runtime supports the feature can be determined by checking presence of `System.Runtime.CompilerServices.RuntimeFeature.ByRefLikeGenerics` field.
The APIs were added in https://github.com/dotnet/runtime/pull/98070.

### `using` statement

A `using` statement will recognize and use implementation of `IDisposable` interface when resource is a ref struct.
```csharp
ref struct S2 : System.IDisposable
{
    void System.IDisposable.Dispose()
    {
    }
}

class C
{
    static void Main()
    {
        using (new S2())
        {
        } // S2.System.IDisposable.Dispose is called
    }
}
```

Note that preference is given to a `Dispose` method that implements the pattern, and only if one is not found, `IDisposable`
implementation is used.

A `using` statement will recognize and use implementation of `IDisposable` interface when resource is a type parameter that 

> `allows ref strict` and `IDisposable` is in its effective interfaces set.

Reviewer 2: ```suggestion
`allows ref struct` and `IDisposable` is in its effective interfaces set.
```

An `await using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a ref struct.
```csharp
ref struct S2 : IAsyncDisposable
{
    ValueTask IAsyncDisposable.DisposeAsync()
    {
    }
}

class C
{
    static async Task Main()
    {
        await using (new S2())
        {
        } // S2.IAsyncDisposable.DisposeAsync
    }
}
```

Note that preference is given to a `DisposeAsync` method that implements the pattern, and only if one is not found, `IAsyncDisposable`
implementation is used.

A pattern `DisposeAsync` method will be recognized on a type parameter that `allows ref struct` as it is recognized on
type parameters without that constraint today.

```csharp
interface IMyAsyncDisposable
{
    ValueTask DisposeAsync();
}

class C
{
    static async Task Test<T>() where T : IMyAsyncDisposable, new(), allows ref struct
    {
        await using (new T())
        {
        } // IMyAsyncDisposable.DisposeAsync
    }
}
```

A `using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a type parameter that 

> `allows ref strict`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.

Reviewer 3: >strict [](http://example.com/codeflow?start=12&length=6)

Typo.

> `allows ref strict`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.

Reviewer 3: ```suggestion
`allows ref struct`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.
```

Reviewer 3: > A `using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a type parameter that 
> `allows ref strict`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.

Currently, it looks like we use `IAsyncDisposable` even if there is a `DisposeAsync` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwKwG5nICWAdgC4CmATgGYCGAxuegJICyAngIIDO7x9AEULcADgHtutAEYAbcoiQBvZOlXoAarRkBXclABs6IaInkefegAoAlPiQBfAihjoAwsmVI1mOIagAOTEMAFXJuUgAeYIA+G3QAdwALKiZg9BAWDnN+Y3FJWXkAGhZswWE86Tli4nJ4mxU1T29vKABOTAxLGvj0YJtrBub0JqHmrAA6VzFibjE5cYB1SkIKSzgYAGZbQeb7dAB6fZLeHPKJSvJx3NNSnccHIA==).

Author: > > A `using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a type parameter that
> > `allows ref strict`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.
> 
> Currently, it looks like we use `IAsyncDisposable` even if there is a `DisposeAsync` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwKwG5nICWAdgC4CmATgGYCGAxuegJICyAngIIDO7x9AEULcADgHtutAEYAbcoiQBvZOlXoAarRkBXclABs6IaInkefegAoAlPiQBfAihjoAwsmVI1mOIagAOTEMAFXJuUgAeYIA+G3QAdwALKiZg9BAWDnN+Y3FJWXkAGhZswWE86Tli4nJ4mxU1T29vKABOTAxLGvj0YJtrBub0JqHmrAA6VzFibjE5cYB1SkIKSzgYAGZbQeb7dAB6fZLeHPKJSvJx3NNSnccHIA==).

I can see how this could be confusing, but there is no pattern match in this case due to an ambiguity.

Reviewer 3: My comment here (and below) was about the text rather than the example. The text states we'll use the interface implementation when there is no pattern implementation, which seems to imply that if there was a pattern implementation, we'd use that instead of the interface implementation. But currently it looks like we use the interface implementation regardless of whether there is a pattern implementation.

Author: > there is no `DisposeAsync` pattern match

The statement doesn't talk about " a pattern implementation" there, it talks about "a pattern match", which doesn't have to exists (and in fact doesn't in some scenarios) when there is a pattern implementation.

            System.Console.Write(123);
        } // IAsyncDisposable.DisposeAsync
    }
}
```

### `foreach` statement

A `foreach` statement will recognize and use implementation of ```IEnumerable<T>```/```IEnumerable``` interface when collection is a ref struct.
```csharp
ref struct S : IEnumerable<int>
{
    IEnumerator<int> IEnumerable<int>.GetEnumerator() {...}
    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() {...}
}

class C
{
    static void Main()
    {
        foreach (var i in new S()) // IEnumerable<int>.GetEnumerator
        {
        }
    }
}
```

A pattern `GetEnumerator` method will be recognized on a type parameter that `allows ref struct` as it is recognized on
type parameters without that constraint today.

```csharp
interface IMyEnumerable<T>
{
    IEnumerator<T> GetEnumerator();
}

class C
{
    static void Test<T>(T t) where T : IMyEnumerable<int>, allows ref struct
    {
        foreach (var i in t) // IMyEnumerable<int>.GetEnumerator
        {
        }
    }
}
```

A `foreach` statement will recognize and use implementation of ```IEnumerable<T>```/```IEnumerable``` interface when collection is a type parameter that 

> `allows ref strict`, there is no `GetEnumerator` pattern match, and ```IEnumerable<T>```/```IEnumerable``` is in type parameter's effective interfaces set.

Reviewer 3: >strict [](http://example.com/codeflow?start=12&length=6)

Typo.

> `allows ref strict`, there is no `GetEnumerator` pattern match, and ```IEnumerable<T>```/```IEnumerable``` is in type parameter's effective interfaces set.

Reviewer 3: ```suggestion
`allows ref struct`, there is no `GetEnumerator` pattern match, and ```IEnumerable<T>```/```IEnumerable``` is in type parameter's effective interfaces set.
```

Reviewer 3: > A `foreach` statement will recognize and use implementation of ```IEnumerable<T>```/```IEnumerable``` interface when collection is a type parameter that 
> `allows ref strict`, there is no `GetEnumerator` pattern match, and ```IEnumerable<T>```/```IEnumerable``` is in type parameter's effective interfaces set.

Currently, it looks like we use `IEnumerable<T>` even if there is a `GetEnumerator` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQWCWA7AFwFMAnAMwEMBjIuASQFkBPAUTwFcBbUigIwBsiiADwAVAHxYA3ljhz6bLjwIB7EsPwFxcAOJECi7iQqqSACgCUGTAF8sWKACY4AYWmz5SAGwJkcUUQAzgRi4maicAQWcADuABaktBEg9MyGPAJCGoTiADQIiADM2VoecjKY8lVwZGpE1HFwZgBuFCRwOB14kdEA9L0FxZriAHR6BhxGJmpl1RXV8naVi1g2QA).

Author: > > A `foreach` statement will recognize and use implementation of `IEnumerable<T>`/`IEnumerable` interface when collection is a type parameter that
> > `allows ref strict`, there is no `GetEnumerator` pattern match, and `IEnumerable<T>`/`IEnumerable` is in type parameter's effective interfaces set.
> 
> Currently, it looks like we use `IEnumerable<T>` even if there is a `GetEnumerator` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQWCWA7AFwFMAnAMwEMBjIuASQFkBPAUTwFcBbUigIwBsiiADwAVAHxYA3ljhz6bLjwIB7EsPwFxcAOJECi7iQqqSACgCUGTAF8sWKACY4AYWmz5SAGwJkcUUQAzgRi4maicAQWcADuABaktBEg9MyGPAJCGoTiADQIiADM2VoecjKY8lVwZGpE1HFwZgBuFCRwOB14kdEA9L0FxZriAHR6BhxGJmpl1RXV8naVi1g2QA).

I can see how this could be confusing, but there is no pattern match in this case due to an ambiguity.

```csharp
interface IMyEnumerable1<T>
{
    IEnumerator<int> GetEnumerator();
}

interface IMyEnumerable2<T>
{
    IEnumerator<int> GetEnumerator();
}

class C
{
    static void Test<T>(T t) where T : IMyEnumerable1<int>, IMyEnumerable2<int>, IEnumerable<int>, allows ref struct
    {
        foreach (var i in t) // IEnumerable<int>.GetEnumerator
        {
        }
    }
}
```

> An `enumarator` pattern will be recognized on a type parameter that `allows ref struct` as it is recognized on

Reviewer 3: ```suggestion
An `enumerator` pattern will be recognized on a type parameter that `allows ref struct` as it is recognized on
```

type parameters without that constraint today.

```csharp
interface IGetEnumerator<TEnumerator> where TEnumerator : allows ref struct 
{
    TEnumerator GetEnumerator();
}

class C
{
    static void Test1<TEnumerable, TEnumerator>(TEnumerable t)
        where TEnumerable : IGetEnumerator<TEnumerator>, allows ref struct
        where TEnumerator : IEnumerator, IDisposable, allows ref struct 
    {
        foreach (var i in t) // IEnumerator.MoveNext/Current
        {
        }
    }

    static void Test2<TEnumerable, TEnumerator>(TEnumerable t)
        where TEnumerable : IGetEnumerator<TEnumerator>, allows ref struct
        where TEnumerator : IEnumerator<int>, allows ref struct 
    {
        foreach (var i in t) // IEnumerator<int>.MoveNext/Current
        {
        }
    }

>     static void Test2<TEnumerable, TEnumerator>(TEnumerable t)

Reviewer 3: ```suggestion
    static void Test3<TEnumerable, TEnumerator>(TEnumerable t)
```

        where TEnumerable : IGetEnumerator<TEnumerator>, allows ref struct
        where TEnumerator : IMyEnumerator<int>, allows ref struct 
    {
        foreach (var i in t) // IMyEnumerator<int>.MoveNext/Current
        {
        }
    }
}

interface IMyEnumerator<T> : System.IDisposable
{
    T Current {get;}
    bool MoveNext();
}
```

A `foreach` statement will recognize and use implementation of `IDisposable` interface when enumerator is a ref struct.
```csharp
struct S1
{
    public S2 GetEnumerator()
    {
        return new S2();
    }
}

ref struct S2 : System.IDisposable
{
    public int Current {...}
    public bool MoveNext() {...}
    void System.IDisposable.Dispose() {...}
}

class C
{
    static void Main()
    {
        foreach (var i in new S1())
        {
        } // S2.System.IDisposable.Dispose()
    }
}
```

Note that preference is given to a `Dispose` method that implements the pattern, and only if one is not found, `IDisposable`
implementation is used.

A `foreach` statement will recognize and use implementation of `IDisposable` interface when enumerator is a type parameter that 

> `allows ref strict` and `IDisposable` is in its effective interfaces set.

Reviewer 3: >strict [](http://example.com/codeflow?start=12&length=6)

Typo.

> `allows ref strict` and `IDisposable` is in its effective interfaces set.

Reviewer 3: ```suggestion
`allows ref struct` and `IDisposable` is in its effective interfaces set.
```

        }
    }
}
```

### `await foreach` statement

An `await foreach` statement will recognize and use implementation of ```IAsyncEnumerable<T>``` interface when collection is a ref struct.
```csharp
ref struct S : IAsyncEnumerable<int>
{
    IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(CancellationToken token) {...}
}

class C
{
    static async Task Main()
    {
        await foreach (var i in new S()) // S.IAsyncEnumerable<int>.GetAsyncEnumerator
        {
        }
    }
}
```

A pattern `GetAsyncEnumerator` method will be recognized on a type parameter that `allows ref struct` as it is recognized on
type parameters without that constraint today.

```csharp
interface IMyAsyncEnumerable<T>
{
    IAsyncEnumerator<int> GetAsyncEnumerator(CancellationToken cancellationToken = default);
}

class C
{
    static async Task Test<T>() where T : IMyAsyncEnumerable<int>, allows ref struct
    {
        await foreach (var i in default(T)) // IMyAsyncEnumerable<int>.GetAsyncEnumerator
        {
        }
    }
}
```

An `await foreach` statement will recognize and use implementation of ```IAsyncEnumerable<T>``` interface when collection is a type parameter that 

> `allows ref strict`, there is no `GetAsyncEnumerator` pattern match, and ```IAsyncEnumerable<T>``` is in type parameter's effective interfaces set.

Reviewer 3: >strict [](http://example.com/codeflow?start=12&length=6)

Typo.

> `allows ref strict`, there is no `GetAsyncEnumerator` pattern match, and ```IAsyncEnumerable<T>``` is in type parameter's effective interfaces set.

Reviewer 3: ```suggestion
`allows ref struct`, there is no `GetAsyncEnumerator` pattern match, and ```IAsyncEnumerable<T>``` is in type parameter's effective interfaces set.
```

Reviewer 3: > An `await foreach` statement will recognize and use implementation of ```IAsyncEnumerable<T>``` interface when collection is a type parameter that 
> `allows ref strict`, there is no `GetAsyncEnumerator` pattern match, and ```IAsyncEnumerable<T>``` is in type parameter's effective interfaces set.

Currently, `IAsyncEnumerable<T>` is used even if there is a `GetAsyncEnumerator` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5lqZwB0AKgBYBOApgIbACWAdgOb4oZYCs7yzALtUoAzWgGNq6AJIBZAJ4BBAM5ymYgKJMArgFshtAEYAbanAA8pAHzIA3snQPpy1Ru17KtfgHtKZgZfQAcWp+ZzVNXX1vSgAKAGFaNWojI08GLyZSLwBraiZ0MUSJFLSMrNz8gF50YGpRLSN+AEp2AF9kAhh0ONt7RywANkwADkwh0molfgtLGKb0AHdyIUlSdBBpeTDXSI9jUz8mfksAGicVcLd9fcPjvoc7JEdnzABOdGEfOjFydBiAN1olHQDBB+Vq9UaMVITXmAHo4ecXBF3IYTLdLMRgqELjtUdF7i9Hi8SURiHEMkovCZiAB1SgMQQxBgtQnPdpPRwc1pAA===).

Author: > > An `await foreach` statement will recognize and use implementation of `IAsyncEnumerable<T>` interface when collection is a type parameter that
> > `allows ref strict`, there is no `GetAsyncEnumerator` pattern match, and `IAsyncEnumerable<T>` is in type parameter's effective interfaces set.
> 
> Currently, `IAsyncEnumerable<T>` is used even if there is a `GetAsyncEnumerator` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5lqZwB0AKgBYBOApgIbACWAdgOb4oZYCs7yzALtUoAzWgGNq6AJIBZAJ4BBAM5ymYgKJMArgFshtAEYAbanAA8pAHzIA3snQPpy1Ru17KtfgHtKZgZfQAcWp+ZzVNXX1vSgAKAGFaNWojI08GLyZSLwBraiZ0MUSJFLSMrNz8gF50YGpRLSN+AEp2AF9kAhh0ONt7RywANkwADkwh0molfgtLGKb0AHdyIUlSdBBpeTDXSI9jUz8mfksAGicVcLd9fcPjvoc7JEdnzABOdGEfOjFydBiAN1olHQDBB+Vq9UaMVITXmAHo4ecXBF3IYTLdLMRgqELjtUdF7i9Hi8SURiHEMkovCZiAB1SgMQQxBgtQnPdpPRwc1pAA===).

I can see how this could be confusing, but there is no pattern match in this case due to an ambiguity.
